### PR TITLE
Support env var parameter substitution

### PR DESCRIPTION
**Pull request by solimant**
## Hey, I just made a Pull Request!

This PR adds the support for parameter substitution in environment variables. For example, `foo ${MISSING:-default-value}` becomes `foo default-value` if `MISSING` is an unset environment variable, or is declared but has no value. This is useful in situations where an environment variable is defined in `app-config.yaml` (e.g. in a proxy block), but a fallback value is desired when the environment variable is not set. For example:
```yaml
proxy:
  endpoints:
    '/foo':
      headers:
        authorization: 'token ${FOO_TOKEN:-my-local-dev-token}'
      target: ${FOO_PROXY_TARGET:-http://localhost:9000}
```
When `FOO_TOKEN` and/or `FOO_PROXY_TARGET` are unset, they will be substituted with `my-local-dev-token` and `http://localhost:9000`, respectively.

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [X] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by solimant** [comment]
@freben, @tudi2d, @Rugvip - does this look good to you?

**Comment by freben** [comment]
One more note. This needs the docs to be updated in https://backstage.io/docs/conf/writing#environment-variable-substitution as part of the PR.

Adding @Rugvip for info too

File: packages/config-loader/src/sources/transform/substitution.ts [hunk 1]
```diff
@@ -37,7 +39,16 @@ export function createSubstitutionTransform(
       if (part.startsWith('$$')) {
         parts[i] = part.slice(1);
       } else {
-        parts[i] = await env(part.slice(2, -1).trim());
+        if (part.includes(':-')) {
+          const subparts = part.split(':-', 2);
```
[the remark this answers is not in the document]
Comment (solimant) [reply packages/config-loader/src/sources/transform/substitution.ts]: Good points 👍🏼 Done.

**Comment by solimant** [comment]
@freben - hey Fredrik, thank you for the review! I've addressed your comments, and also have updated the docs. Happy to address any further comments.

File: docs/conf/writing.md [hunk 2]
```diff
@@ -199,6 +199,18 @@ configuration value will evaluate to `undefined`.
 The substitution syntax can be escaped using `$${...}`, which will be resolved
 as `${...}`.
 
+Parameter substitution syntax (e.g. `${MY_VAR:-default-value}`) is also
+supported to provide a default or fallback value for a given environment
+variable if it is unset, or is declared but has no value. For example:
+
+```yaml
+app:
+  baseUrl: https://${HOST:-http://localhost:3000}
```
Comment (Rugvip) [inline docs/conf/writing.md:208]:
```suggestion
app:
  baseUrl: ${HOST:-https://localhost:3000}
```

Something like this instead right?

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you for the suggestions! 👍 

Wanna perhaps push back a little bit on this as it has quite a bit of overlap with the `APP_CONFIG_*` env vars. It's ofc doesn't quite handle the case of substitution within a value, but I'm not quite sure how much of a problem that is in practice.

The `proxy.endpoints` isn't really something that we'd want to be designing the config system around tbh. Long-term it's something that I hope we're able to redesign into something that provides more out of the box and is more secure. Do you happen to have any other other example where this might be useful?

**Comment by solimant** [comment]
> Thank you for the suggestions! 👍
> 
> Wanna perhaps push back a little bit on this as it has quite a bit of overlap with the `APP_CONFIG_*` env vars. It's ofc doesn't quite handle the case of substitution within a value, but I'm not quite sure how much of a problem that is in practice.
> 
> The `proxy.endpoints` isn't really something that we'd want to be designing the config system around tbh. Long-term it's something that I hope we're able to redesign into something that provides more out of the box and is more secure. Do you happen to have any other other example where this might be useful?

Thank you, Patrik. Yeah, in practice, we want to be able to define env vars in `app-config.yaml`, while reducing the amount of friction in asking our developers to declare them in their local dev environments, or introduce scripts to set them. Keeping the env var fallback values in `app-config.yaml` helps with that. Also, specifically for the proxy endpoints use-case, the `APP_CONFIG_*` override can't handle that use-case. This means that you'd need to do something like this:

```yaml
proxy:
  endpoints:
    '/foo':
      target: ${PROXY_FOO}
```
In such use-case, you'll have to set `PROXY_FOO` somehow for local dev environment via a script. Otherwise, it will break, especially when people forget to update that script, which creates friction.

Similar use-cases are when we're dealing with array-like configurations; for example:
```yaml
catalog:
  locations:
    - type: file
      target: ${CATALOG_INFO_YAML}
      rules:
        - allow: [Component, Resource, User, Group]
```
`APP_CONFIG_*` can't handle that use-case as well; i.e. something like `APP_CONFIG_catalog_locations_0_target` isn't supported, and even if it was, it won't work nicely with arrays because you'll need to target the array element by its index, which can become annoying and hard to maintain.

So, with parameter substitution, use-cases can be addressed with less code on the developer's end:
```yaml
proxy:
  endpoints:
    '/foo':
      target: ${PROXY_FOO:-http://dev.foo.test}

catalog:
  locations:
    - type: file
      target: ${CATALOG_INFO_YAML:-local-catalog-info.yaml}
```

**Review by Rugvip — APPROVED** [review]
Thank you! 👍 

Let's :shipit:, this is standard enough and close enough to our current capabilities that I'm not seeing too much risk wrt complexity.

**Pull request merged**